lib/Dialect/HW/Transforms/HWAggregateToComb.cpp

```diff
@@ -0,0 +1,186 @@
+//===- HWAggregateToComb.cpp - HW aggregate to comb -------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWPasses.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace circt {
+namespace hw {
+#define GEN_PASS_DEF_HWAGGREGATETOCOMB
+#include "circt/Dialect/HW/Passes.h.inc"
+} // namespace hw
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+
+namespace {
+template <typename OpTy>
+struct HWArrayCreateLikeOpConversion : OpConversionPattern<OpTy> {
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  using OpAdaptor = typename OpConversionPattern<OpTy>::OpAdaptor;
+  LogicalResult
+  matchAndRewrite(OpTy op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Lower to concat.
+    auto inputs = adaptor.getInputs();
+    SmallVector<Value> results;
+    for (auto input : inputs)
+      results.push_back(rewriter.getRemappedValue(input));
+    rewriter.replaceOpWithNewOp<comb::ConcatOp>(op, results);
+    return success();
+  }
+};
+
+struct HWAggregateConstantOpConversion
+    : OpConversionPattern<hw::AggregateConstantOp> {
+  using OpConversionPattern<hw::AggregateConstantOp>::OpConversionPattern;
+
+  static LogicalResult peelAttribute(Location loc, Attribute attr,
+                                     ConversionPatternRewriter &rewriter,
+                                     SmallVector<Value> &results) {
+    SmallVector<Attribute> worklist;
+    worklist.push_back(attr);
+
+    while (!worklist.empty()) {
+      auto current = worklist.pop_back_val();
+      if (auto innerArray = dyn_cast<ArrayAttr>(current)) {
+        for (auto elem : llvm::reverse(innerArray))
+          worklist.push_back(elem);
+        continue;
+      }
+
+      if (auto intAttr = dyn_cast<IntegerAttr>(current)) {
+        results.push_back(rewriter.create<hw::ConstantOp>(loc, intAttr));
```

Comment:
Wouldn't it be better to have an APInt and using its `concat` method rather than materializing all this IR just to fold it?